pipelines/controller/src/main/java/com/google/fhir/analytics/FlinkConfiguration.java

```diff
@@ -187,6 +189,51 @@ private void deleteFilesRecursivelyOnExit(Path path) {
                 }));
   }
 
+  /**
+   * Validates if the JVM memory is sufficient to run pipelines successfully. This validation helps
+   * the application to fail-fast rather than failing during batch execution.
+   */
+  private void validateHeapMemoryConfig(DataProperties dataProperties) {
+    long maxMemory = Runtime.getRuntime().maxMemory();
+    long minJVMMemoryRequired = minJVMMemory(dataProperties);
+    if (minJVMMemoryRequired > maxMemory) {
+      throw new IllegalConfigurationException(
+          String.format(
+              "Insufficient max JVM memory, required %s mb but provided %s mb",
+              memoryInMB(minJVMMemoryRequired), memoryInMB(maxMemory)));
+    }
+  }
+
+  /**
+   * The minimum JVM memory needed for the pipelines to run without fail. This is derived based on
+   * the number of parquet row groups that can be read or written in parallel by the pipeline
+   * threads. This is roughly derived by the below formula.
+   *
+   * <p>Minimum JVM Memory = Memory for JVM stack, perm files etc + (#Parallel Pipeline Threads *
+   * #Parallel Pipelines * (Parquet Row Group Size + buffer for parsing the parquet file))
+   */
+  private long minJVMMemory(DataProperties dataProperties) {
+    int rowGroupSize =
+        dataProperties.getRowGroupSizeForParquetFiles() > 0
+            ? dataProperties.getRowGroupSizeForParquetFiles()
```

Comment:
nit: The only caveat here is that if the files are written using a higher row-group size (e.g., with one of the previous releases of the pipeline), now reading them back into memory needs more memory (since `getRowGroupSizeForParquetFiles` is for future writes). I don't think we need to make any code changes for this but it is probably good to document this somewhere.

Reply:
Yes, this make sense and I have documented the same.